src/Npgsql/MultiHostConnectorPool.cs

```diff
@@ -0,0 +1,261 @@
+                var portSeparator = host.IndexOf(':');
+                if (portSeparator != -1)
+                {
+                    host = host.Substring(0, portSeparator);
+                    port = int.Parse(host.Substring(portSeparator + 1));
+                }
+                var poolSettings = settings.Clone();
+                poolSettings.Host = host;
+                poolSettings.Port = port;
+                _pools[i] = new ConnectorPool(poolSettings, poolSettings.ConnectionString, this);
+                // TODO: should be safely removed
+                /*
+                // Setting the default settings due to PoolManager.TryGetValue
+                _pools[i] = new ConnectorPool(poolSettings, settings.ConnectionString, this);
+                */
+            }
+        }
+
+        static bool IsValidHost(ConnectorPool pool, TargetServerType preferedType)
+        {
+            var currentState = GetHostState(pool);
+            if (currentState == HostState.Offline)
+                return false;
+            return preferedType == TargetServerType.Any || currentState == HostState.Unknown ||
+                preferedType.HasFlag(TargetServerType.Primary) && currentState == HostState.Primary ||
+                preferedType.HasFlag(TargetServerType.Secondary) && currentState == HostState.Secondary;
+        }
+
+        static bool IsValidUnpreferedHost(ConnectorPool pool, TargetServerType preferedType)
+        {
+            var currentState = GetHostState(pool);
+            if (currentState == HostState.Offline)
+                return false;
+            return currentState == HostState.Unknown ||
+                preferedType == TargetServerType.PreferSecondary && currentState == HostState.Primary ||
+                preferedType == TargetServerType.PreferPrimary && currentState == HostState.Secondary;
+        }
+
+        static HostState GetHostState(ConnectorPool pool)
+            => HostsCache.GetHostState($"{pool.Settings.Host}:{pool.Settings.Port}");
+
+        async ValueTask<NpgsqlConnector?> TryGetIdle(NpgsqlConnection conn, TimeSpan timeoutPerHost, bool async, TargetServerType preferedType,
+            Func<ConnectorPool, TargetServerType, bool> validHostFunc, IList<Exception> exceptions,
+            CancellationToken cancellationToken)
+        {
+            foreach (var pool in _pools)
+            {
+                if (!validHostFunc(pool, preferedType))
+                    continue;
+
+                if (pool.TryGetIdleConnector(out var connector))
+                {
+                    conn.Connector = connector;
+                    connector.Connection = conn;
+
+                    try
+                    {
+                        if (GetHostState(pool) == HostState.Unknown)
+                            await connector.UpdateServerType(new NpgsqlTimeout(timeoutPerHost), async, cancellationToken);
+
+                        if (!validHostFunc(pool, preferedType))
```

Comment:
If I'm reading this right, the cluster state can only change if it was Unknown above, and we called UpdateServerType, right? If so, this can go into the condition above?

Reply:
Yeah, I think you're correct and we can do this.